### PR TITLE
Fix mask detection

### PR DESCRIPTION
**Pull request by mcasperson**
The detection of a mask was incorrect, allowing any version with a prelease field that ended in `i` or `c` to be used as a mask.

This PR fixes the detection of the mask to only match versions where the prerelease ends with a period and a `i` or `c`, which was the original logic [here](https://github.com/OctopusDeploy/OctopusDeploy/blob/43e42a5769083dffa65e2c9ab4d513641c37248a/source/Octopus.Core/Resources/SemanticVersionMask.cs#L217).

This PR also includes a bunch of tests to validate the change.

Fixes https://github.com/OctopusDeploy/Issues/issues/6926

### CODE REVIEW

**Comment by mcasperson** [comment]
[Version masks are incorrectly applied when the prerelease field ends with a i or c.](https://trello.com/c/LfP1XFBu/106-version-masks-are-incorrectly-applied-when-the-prerelease-field-ends-with-a-i-or-c)

**Review by iryan2 — APPROVED** [review]
Changes look good to me, I've also run the tests locally 👍

File: source/Octopus.Versioning/Octopus/OctopusVersionMask.cs [hunk 1]
```diff
@@ -177,7 +177,7 @@ public override bool IsSubstitute
                     if (!IsPresent)
                         return false;
 
-                    return Regex.IsMatch(matchGroup.Value, @$"{OctopusVersionMaskParser.PatternIncrement}|{OctopusVersionMaskParser.PatternCurrent}$");
+                    return Regex.IsMatch(matchGroup.Value, @$"\.({OctopusVersionMaskParser.PatternIncrement}|{OctopusVersionMaskParser.PatternCurrent})$");
```
Comment (rhysparry) [inline source/Octopus.Versioning/Octopus/OctopusVersionMask.cs:180]:
```suggestion
                    return Regex.IsMatch(matchGroup.Value, @$"\.(?:{OctopusVersionMaskParser.PatternIncrement}|{OctopusVersionMaskParser.PatternCurrent})$");
```

If not capturing a group, you should explicitly mark it as non-capturing.

This was the result of a simple test I ran against this code:

|       Method |     Mean |    Error |  StdDev |
|------------- |---------:|---------:|--------:|
|    Capturing | 372.6 ns | 18.68 ns | 1.02 ns |
| NonCapturing | 335.9 ns | 12.76 ns | 0.70 ns |

Trivial for this example, but a good practice nevertheless.

Comment (mcasperson) [reply source/Octopus.Versioning/Octopus/OctopusVersionMask.cs]: 👍 I didn't realise that had a performance benefit. I'll add that change.

**Pull request merged**